### PR TITLE
Changes PostCompletion to handle OnAsyncCompletion after managed request has completed. 

### PR DESCRIPTION
**Pull request by jkotalik**
More stuff for the Async PR in IISIntegration: https://github.com/aspnet/IISIntegration/pull/447

### CODE REVIEW

File: src/AspNetCore/Src/managedexports.cxx [hunk 1]
```diff
@@ -56,9 +56,24 @@ EXTERN_C __MIDL_DECLSPEC_DLLEXPORT
 HRESULT
 http_post_completion(
```
Comment (davidfowl) [inline src/AspNetCore/Src/managedexports.cxx]: I don't think this logic should be in post_completion

**Comment by jkotalik** [comment]
@pan-wang We are going to be calling PostCompletion from managed to native. By calling this, OnAsyncCompletion will be triggered again by the proxymodule, which cannot call back into managed as managed will be disposing. To handle this, before calling post completion, we call http_set_completion_status() to set the status and that we have completed processing in managed. We then check in PostCompletion if the request has completed, and if so we return the status.

**Comment by jkotalik** [comment]
This needs to be synced with https://github.com/aspnet/IISIntegration/pull/447

File: src/AspNetCore/Src/inprocessapplication.cxx [hunk 2]
```diff
@@ -38,9 +39,22 @@ IN_PROCESS_APPLICATION::OnAsyncCompletion(
...
-    // Call the managed handler for async completion.
-    return m_AsyncCompletionHandler(pInProcessStoredContext->QueryManagedHttpContext(), hrCompletionStatus, cbCompletion);
+    else if (pInProcessStoredContext->QueryIsManagedRequestComplete())
+    {
+        // means PostCompletion has been called and this is the associated callback.
+        dwRequestNotificationStatus = pInProcessStoredContext->QueryAsyncCompletionStatus();
+        if (dwRequestNotificationStatus == RQ_NOTIFICATION_FINISH_REQUEST)
+        {
+            pHttpContext->DisableNotifications(RQ_END_REQUEST, 0);
+            // TODO cleanup whatever disconnect listener there is
```
Comment (davidfowl) [inline src/AspNetCore/Src/inprocessapplication.cxx]: Is there an issue for this?

File: src/AspNetCore/Src/inprocessapplication.cxx [hunk 3]
```diff
@@ -38,9 +39,22 @@ IN_PROCESS_APPLICATION::OnAsyncCompletion(
...
-
-    // Call the managed handler for async completion.
-    return m_AsyncCompletionHandler(pInProcessStoredContext->QueryManagedHttpContext(), hrCompletionStatus, cbCompletion);
+    else if (pInProcessStoredContext->QueryIsManagedRequestComplete())
+    {
+        // means PostCompletion has been called and this is the associated callback.
+        dwRequestNotificationStatus = pInProcessStoredContext->QueryAsyncCompletionStatus();
+        if (dwRequestNotificationStatus == RQ_NOTIFICATION_FINISH_REQUEST)
+        {
+            pHttpContext->DisableNotifications(RQ_END_REQUEST, 0);
```
Comment (davidfowl) [inline src/AspNetCore/Src/inprocessapplication.cxx]: do we need this?

Comment (pan-wang) [reply src/AspNetCore/Src/inprocessapplication.cxx]: ANCM will not receive RQ_NOTIFICATION_FINISH_REQUEST as it did not register it.
I don't think we need it. Inprocess should always call m_AsyncCompletionHandler

File: src/AspNetCore/Src/inprocessstoredcontext.cxx [hunk 4]
```diff
@@ -32,6 +34,38 @@ IN_PROCESS_STORED_CONTEXT::QueryHttpContext(
     return m_pHttpContext;
 }
 
+BOOL
+IN_PROCESS_STORED_CONTEXT::QueryIsManagedRequestComplete(
```
Comment (pan-wang) [inline src/AspNetCore/Src/inprocessstoredcontext.cxx:38]: the native code ideally should not touch this stored managed context. Not sure why you need it

File: src/AspNetCore/Src/inprocessapplication.cxx (hunk 3 above)
Comment (jkotalik) [reply src/AspNetCore/Src/inprocessapplication.cxx]: Alright I'll drop it.

File: src/AspNetCore/Src/inprocessapplication.cxx (hunk 2 above)
Comment (jkotalik) [reply src/AspNetCore/Src/inprocessapplication.cxx]: I'll make a separate issue for adding the disconnect listener to managed and native.

File: src/AspNetCore/Src/inprocessstoredcontext.cxx (hunk 4 above)
Comment (jkotalik) [reply src/AspNetCore/Src/inprocessstoredcontext.cxx:38]: This doesn't touch the stored context. This is for knowing that an OnAsyncCompletion call was associated with PostCompletion(). If the context calls post completion, we should expect that the request is done and the managed request is being disposed.

**Comment by jkotalik** [comment]
@pan-wang Is this one good to go after offline feedback?

**Review by pan-wang — APPROVED** [review]

**Pull request merged**